kyo-stm/shared/src/main/scala/kyo/TRef.scala

```diff
@@ -0,0 +1,190 @@
+                                val entry = Write(state.tid, v)
+                                Var.setDiscard(log.put(this, entry))
+                            end if
+                        }
+                    }
+        }
+
+    @tailrec private[kyo] def lock(entry: Entry[A])(using AllowUnsafe): Boolean =
+        currentState.tid == entry.tid && {
+            val lockState = super.get()
+            entry match
+                case Read(tid, value) =>
+                    // Read locks can stack if no write lock
+                    lockState != Int.MaxValue && (super.compareAndSet(lockState, lockState + 1) || lock(entry))
+                case Write(tid, value) =>
+                    // Write lock requires no existing locks
+                    lockState == 0 && (super.compareAndSet(lockState, Int.MaxValue) || lock(entry))
+            end match
+        }
+
+    private[kyo] def commit(tid: Long, entry: Entry[A])(using AllowUnsafe): Unit =
+        entry match
+            case Write(_, value) =>
+                // Only need to commit Write entries
+                currentState = Write(tid, value)
+            case _ =>
+
+    private[kyo] def unlock(entry: Entry[A])(using AllowUnsafe): Unit =
+        entry match
+            case Read(tid, value) =>
+                // Release read lock
+                discard(super.decrementAndGet())
+            case Write(tid, value) =>
+                // Release write lock
+                super.set(0)
+        end match
+    end unlock
+end TRefImpl
+
+object TRef:
+
+    /** Creates a new transactional reference within an STM transaction.
+      *
+      * @param value
+      *   The initial value for the reference
+      * @return
+      *   A new transactional reference containing the value, within the STM effect
+      */
+    def init[A](value: A)(using Frame): TRef[A] < STM =
+        useRequiredTid { tid =>
+            Var.use[RefLog] { log =>
+                IO.Unsafe {
+                    val ref    = TRef.Unsafe.init(tid, value)
+                    val refAny = ref.asInstanceOf[TRef[Any]]
+                    Var.setAndThen(log.put(refAny, refAny.state))(ref)
+                }
+            }
+        }
+
+    /** Creates a new transactional reference outside of any transaction.
+      *
+      * WARNING: This operation:
+      *   - Cannot be rolled back
+      *   - Is not part of any transaction
+      *   - Will cause any containing transaction to retry if used within one, since it creates a reference with a newer transaction ID
+      *
+      * Use this only for static initialization or when you specifically need non-transactional creation. For most cases, prefer `init`.
+      *
+      * @param value
+      *   The initial value for the reference
+      * @return
+      *   A new transactional reference containing the value, within the IO effect
+      */
+    def initNow[A](value: A)(using Frame): TRef[A] < IO =
+        IO.Unsafe(TRef.Unsafe.initNow(value))
```

Comment:
I'm not yet sure how important this is to expose outside of the `Unsafe` API.

Reply:
It's useful in regular safe code as well. The tests have examples of a ref being initialized first outside of a transaction and then the following transactions using it.